Remove duplicated Env::Default to fix gcs issue on Windows

yongtang opened the pull request:
This PR tries to address the issue raised in #49515 where gcs is not available through tf.io.gfile.GFile interface.

The reason was that on Windows -config=monolithic was used in bazel which includes multiple copies of `Env::Default` in different pyd files.

This PR removes all duplications to make gcs working on Windows.

This PR fixes #49515

/cc @mihaimaruseac @vnvo2409 @kvignesh1420 @terrytangyuan @burgerkingeater FYI

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Remove duplicated `Env::Default` to fix gcs issue on Windows

This PR tries to address the issue raised in 49515 where
gcs is not available through tf.io.gfile.GFile interface.

The reason was that on Windows -config=monolithic was used
in bazel which includes multiple copies of `Env::Default`
in different pyd files.

This PR removes all duplications to make gcs working on Windows.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

terrytangyuan reviewed (approved):
LGTM

yongtang committed:
Add missing dependency by using tf_python_pybind_extension

and add missing linkage dependency

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add Env::Default to windows def file

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

mihaimaruseac reviewed (approved):
This is a great fix, thank you for it.

mihaimaruseac commented on tensorflow/python/lib/io/file_io_integration_test.py:
```diff
@@ -0,0 +1,35 @@
+# limitations under the License.
+# ==============================================================================
+
+import tensorflow as tf
```
Is there a way to change this to not import TF fully? As it is written this raises this error (sanitized):

```
HOURGLASS IMPORT DETECTED

'import tensorflow' statements are no longer allowed in the internal
TensorFlow codebase, with the exception of /examples/ and /g3doc/
directories. This is because they prevent Bazel and <testing tool> from caching
anything.

Please import modules directly. Also please make sure your Python
BUILD rules depend on all the labels whose sources they import.
```

yongtang replied:
@mihaimaruseac Let me take a look. I think it should be possible. The reason was to enable all modules (so that in case multiple copies of `Env::Default` is present in the future again, it will trigger the test fail). I will see if I can find an internal import that can achieve the same thing.

mihaimaruseac replied:
Alternatively, maybe we can change [the nightly smoke test](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/tools/ci_build/builds/nightly_release_smoke_test.sh;l=24;drc=82f83aae97ae11736e577fa560e9d33cf764a018) to test for this instead. It will prevent nightlies from being pushed and will still give a signal.

yongtang replied:
Thanks @mihaimaruseac . Let me see if I can update the PR.

yongtang committed:
Add test case in nightly release smoke test to make sure gcs is available on Windows through tf.io.gfile.GFile

See GitHub issue #49515 for details.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @mihaimaruseac for the review. I have updated the PR to change the test to smoke test in:
`tensorflow/tools/ci_build/builds/nightly_release_smoke_test.sh`

Please take a look and let me know if this works.

gbaned commented:
@yongtang  Can you please resolve conflicts? Thanks!

yongtang commented:
@mihaimaruseac @gbaned Thanks for the help. The PR has been rebased to resolve the merge conflict. Can you try re-import again?

mihaimaruseac reviewed (approved):
(this needs manual import due to internal filesystem stuff)

mihaimaruseac commented:
This has a large number of internal targets that fail to link due to `Env::Default` no longer being available at link time.

Needs me to go through all these BUILD targets and update

mihaimaruseac commented:
Can we find a solution that does not require external parties to depend on `//tensorflow/core/platform:env_impl`?

It seems even downstream project that want to use `Env::Default()` now need to depend on this internal target.

There are around 1000 targets that need to change internally and this likely won't pass, sadly.

yongtang commented:
@mihaimaruseac Thanks for the update. I will take a look and see if there are any ways to avoid the downstream projects change.

gbaned commented:
@yongtang  Any update on this PR? Please. Thanks!

yongtang commented:
Thanks @gbaned for the reminder. Still working on it as last two weeks has been mostly filled with PTOs and some other activities. I will try to get the PR updated really soon.

yongtang committed:
Use if_static to avoid breaking existing code (for Env::Default duplication fix)

This commit use if_static to avoid breaking existing code for
Env::Default duplication fix. In case of static build (Windows monolithic)
one copy of Env::Default is ensured; in case of non-static build (Linux)
then old behavior is maintained.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix duplicate dependencies

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

mihaimaruseac commented:
Manual import seems needed

yongtang committed:
Fix merge conflict

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

spamz23 commented:
Any update on this? I need this PR to be merged

mihaimaruseac commented:
Will probably need to wait for `cc_shared_library` support given that it breaks internal builds in the current form

yongtang commented:
Thanks @mihaimaruseac . I think cc_shared_library will also resolve #53234 (grpc upgrade issue) as well.

gbaned commented:
Hi @yongtang Can you please resolve conflicts? Thank you!

gbaned commented:
Hi @yongtang Can you please resolve conflicts? Thank you!

gbaned commented:
@yongtang Any update on this PR? Please. Thank you!

gbaned commented:
@yongtang I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen! thank you for your contribution.
